### PR TITLE
Issue #2046 Close connections on non graceful stop

### PR DESCRIPTION
**Pull request by gregw**
Issue #2046 Close connections on non graceful stop
Simplified the ManagedSelector Action mechanism
Ensure that the CloseAction is always run in doStop

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

**Comment by gregw** [comment]
Need to run close even if timeout exception is thrown

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java [hunk 1]
```diff
@@ -115,28 +115,67 @@ protected void doStop() throws Exception
...
+            try
+            {
+                super.doStop();
+            }
+            finally
+            {
+                try (Locker.Lock lock = _locker.lock())
+                {
+                    if (_selector!=null && _selector.isOpen())
+                        _selector.wakeup();
```
Comment (sbordet) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: If above you return a boolean, then you don't need these 2 lines above.
If the boolean is true, you know the selector has been closed, otherwise there is no need to wake it up (it has already been woken up by the submission of the `CloseAction`)

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java [hunk 2]
```diff
@@ -115,28 +115,67 @@ protected void doStop() throws Exception
...
+
+                    _actions.stream().filter(CloseAction.class::isInstance).forEach(a->a.run());
+                    _actions.clear();
+                }
```
Comment (sbordet) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: Nope, we are calling `run()` while holding a lock.
We need to copy the actions we want to run, and then run them without the lock held.
Also, `a->a.run()` can be reduced to a method reference.

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java [hunk 3]
```diff
@@ -115,28 +115,67 @@ protected void doStop() throws Exception
...
+            }
+        }
         if (LOG.isDebugEnabled())
             LOG.debug("Stopped {}", this);
     }
 
-    public void submit(Runnable change)
+    @Deprecated
+    public void submit(Runnable action)
```
Comment (sbordet) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: IMHO this method should be package-private or protected, not public.

I would keep it as `submit(Runnable) ` because actions are treated as Runnables anyway (where we poll them), so while I understand the meaning of "you can only submit Actions, not Runnables" in changing the signature, I think it just buys us very little and clutters the code to have 2 copy of the same method.

File: jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java [hunk 4]
```diff
@@ -55,6 +56,166 @@
 @RunWith(AdvancedRunner.class)
 public class GracefulStopTest 
 {
+    /**
+     * Test of standard graceful timeout mechanism when a block request does
+     * not complete
+     * @throws Exception on test failure
+     */
+    @Test
+    public void testSleepNotGraceful() throws Exception
```
Comment (sbordet) [inline jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java]: Can these 3 new tests be parametrized ? They share 99% of the same code, so perhaps it's possible to just call a private method with 2-3 parameters to control its behavior ?

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java [hunk 5]
```diff
@@ -115,28 +115,67 @@ protected void doStop() throws Exception
...
-        close_selector.await(getStopTimeout());
-
-        super.doStop();
         
+        try
+        {
+            CloseAction close = new CloseAction();
+            submit(close);
+            if (getStopTimeout()>0)
+                close.await(getStopTimeout());
```
[the remark this answers is not in the document]
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: if the timeout is <=0 there is no timeout, so I don't want to wait and get a TimeoutException

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java (hunk 1 above)
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: the previous submit is before the call to super.doStop(), so it may wakeup, loop and then select again before we grab the lock at 134.   Once we have the lock we know the state of the ManagedSelector is `!isRunning()`, so another wakeup is needed to make sure the selecting thread sees that.

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java (hunk 2 above)
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: I don't think so.  The actions are all intended to be run by the selector thread and not dispatched to arbitrary user code.  So it is fine for the lock to be held... in fact it is probably good.

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java (hunk 3 above)
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: I think this is needed to avoid the little mess we made by submitting Blocking runnables.   However I agree that the new submit(Action) can be protected.

File: jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java (hunk 4 above)
Comment (gregw) [reply jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java]: sure

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: Hmmm cancel that... await doesn't throw TimeoutException... but it probably should do as I think that is the contract of a graceful shutdown failure..... looking deeper....

**Comment by gregw** [comment]
@sbordet actually I'm not even testing what I thought I was....   fair bit of rework going in....

**Comment by gregw** [comment]
@sbordet I've made a moderately significant refactor.  It now really tests a slow close and still always forces a close on all endPoints

**Comment by gregw** [comment]
On look at this more, I agree actions should not be called with lock held.... stand by....

**Comment by gregw** [comment]
@sbordet please review updated doStop

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java [hunk 6]
```diff
@@ -115,28 +116,91 @@ protected void doStop() throws Exception
...
+                // close all the endpoints directly
+                selector.keys().stream()
+                .filter(SelectionKey::isValid)
+                .map(SelectionKey::attachment)
+                .filter(EndPoint.class::isInstance)
+                .map(EndPoint.class::cast)
+                .filter(EndPoint::isOpen)
+                .forEach(EndPoint::close);
+                selector.close();
+            }
```
Comment (sbordet) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: This whole `finally` block is racy with `GracefulCloseAction`, so if `GracefulCloseAction` takes more than `stopTimeout`, it will still run when the `finally` block runs.
Calls such as `selector.keys()` throw `ClosedSelectorException` (a runtime exception) if the selector is closed, so we will throw an exception while instead the closing has been done successfully already.
Also, the logic is different from `GracefulCloseAction` because there the close was forwarded to the `Connection`.
And it's racy with both `Selector.isOpen()` and `EndPoint.isOpen()`; the latter may be eliminated, perhaps, as `EndPoint.close()` should be idempotent.

Basically the `finally` block is doing the same of `GracefulCloseAction`, so perhaps we just call `run()` on it from the `finally` block, with an atomic guard in `run()` to run it just once: if it did not run it will now and skip it later, if it's already running it won't run again and it will eventually finish. WDYT ?

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java [hunk 7]
```diff
@@ -115,28 +116,91 @@ protected void doStop() throws Exception
...
+        // Grab a copy of the current selector, so we can force close endpoints later
+        Selector selector;
+        try (Locker.Lock lock = _locker.lock())
+        {
+            selector = _selector;
+        }
+
+        boolean graceful = getStopTimeout()>0;
+        boolean closed = selector==null;
+        if (graceful && !closed)
```
Comment (sbordet) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: If `graceful==false` I'm not sure we want to skip the submission of `GracefulCloseAction`. We still want to run that into the selector thread, just wait until it's done, no ?

File: jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java [hunk 8]
```diff
@@ -154,14 +155,23 @@ protected void doStop() throws Exception
...
+                {
+                    stop(l);
+                }
+                catch(Throwable th)
+                {
+                    mex.add(th);
+                }
             }
         }
+        mex.ifExceptionThrow();
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: Should not this change also made to `destroy()` ?

File: jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java [hunk 9]
```diff
@@ -55,6 +63,136 @@
...
+
+        NoopHandler handler = new NoopHandler();
+        server.setHandler(handler);
+
+        server.start();
+        final int port=connector.getLocalPort();
+        Socket client = new Socket("127.0.0.1", port);
+        client.setSoTimeout(10000);
+        client.getOutputStream().write((
+                "GET / HTTP/1.0\r\n"+
```
Comment (sbordet) [inline jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java]: Wouldn't using HTTP/1.0 cause a race to close the connection on the server, since the dispatched thread will try to close it while the call to `server.stop()` below also will try to close it ?
I think it's better to use HTTP/1.1, since you only look for the end of headers below.
Calling `server.stop()` will trigger the close of the connection, and will be the only trigger.

File: jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java [hunk 10]
```diff
@@ -55,6 +63,136 @@
...
+    }
+
+    /**
+     * Test of graceful stop when close is faster than timeout;
+     * @throws Exception on test failure
+     */
+    @Test
+    public void testSlowCloseGraceful() throws Exception
+    {
+        testSlowClose(5000,1000,greaterThan(999L));
```
Comment (sbordet) [inline jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java]: I've seen tests failing with just 1 ms of room. I would use `greaterThan(500L)`.

File: jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java [hunk 11]
```diff
@@ -55,6 +63,136 @@
...
+            server.stop();
+            assertTrue(stopTimeout==0 || stopTimeout>closeWait);  
+        }
+        catch(Exception e)
+        {
+            assertTrue(stopTimeout>0 && stopTimeout<closeWait);                
+        }
+        long stop = System.nanoTime();
+        
+        // No Graceful wait
```
Comment (sbordet) [inline jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java]: Is this comment correct ?

File: jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java [hunk 12]
```diff
@@ -55,6 +63,136 @@
...
+        while(true)
+        {
+            int r = client.getInputStream().read();
+            if (r==-1)
+                break;
+        }
+
+        // onClose Thread interrupted or completed
+        if (stopTimeout>0)
+            assertTrue(closed.await(1000,TimeUnit.MILLISECONDS));
```
Comment (sbordet) [inline jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java]: Don't you always want to check that the `closed` latch has been counted down ?
Even with `stopTimeout==0`, we want to make sure that the connection has been closed.
We should also check that the `selector` is closed, but there is currently no easy way to access it.

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java (hunk 7 above)
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: The contract for `getStopTimeout()<=0` is that `doStop` will not wait, so how can we wait until it is done?

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java (hunk 6 above)
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: The close block is deliberately different to the GracefulCloseAction.  Graceful close can call Connection.close, which may take time and send close handshakes etc. which may block.  None Graceful close is deliberately straight to the endpoint to close the underlying TCP connection regardless of protocol state.

However, I agree with the race concerns and we should protect with something atomic.   This is partially done by calling doStop before the finally block, but is perhaps not rigorous enough.

File: jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java (hunk 12 above)
Comment (gregw) [reply jetty-server/src/test/java/org/eclipse/jetty/server/GracefulStopTest.java]: The connection itself might not be closed if we are not graceful

File: jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java (hunk 8 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: No. A failed component might be fixed and restarted.  destroy is something that is called just before all references to the object are dropped (or to achieve dropping of all references).

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java (hunk 7 above)
Comment (sbordet) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: @gregw so it is the wait that we need to skip, not the submission of the `GracefulStopAction`, no ?

File: jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java (hunk 8 above)
Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: @gregw are you sure ? If the first component that is destroyed throws an exception, all the others will not be destroyed.

Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: this is in doStop, so nothing to do with destroy.  However your point you raise about exceptions from destroy is a good one at it should catch throwable and keep destroying....

Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: @sbordet I modified destroy

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java (hunk 6 above)
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: Added an atomic reference to the selector to prevent the race

File: jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java (hunk 7 above)
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java]: Why submit a task if we are not going to wait for it? we are going to go close the endpoints and selectors so the submitted task would just be in a race... with latest changes it would probably lose the race and just be a noop

**Comment by gregw** [comment]
The problem with making it atomic, is that if the connection close is slow, then we don't do an endpoint close.... which was the whole point!

**Comment by gregw** [comment]
I've made it a bit less racy by only closing the selector in one place, but non graceful close could still be very none graceful!   Perhaps need a bit of face to face review on this one?

**Comment by gregw** [comment]
@sbordet re-review needed

**Comment by gregw** [comment]
@sbordet ping

**Comment by gregw** [comment]
Closing this in favour of #2100
